File: C/28_Tree/BOJ_06_5639_이진검색트리.c
#include <stdio.h>
#include <stdlib.h>
typedef struct TreeNode {
	int data;
	struct TreeNode* left;
	struct TreeNode* right;
} TreeNode;
void postorder(TreeNode* root) {
	if (root) {
		postorder(root->left);
		postorder(root->right);
		printf("%d\n", root->data);
	}
}
TreeNode* new_node(int data) {
	TreeNode* tmp = (TreeNode*)malloc(sizeof(TreeNode));
	tmp->data = data;
	tmp->left = tmp->right = NULL;
	return tmp;
}
TreeNode* insert_node(TreeNode* node, int data) {
	if (node == NULL) return new_node(data);
	if (data < node->data) {
		node->left = insert_node(node->left, data);
	}
	else if (data > node->data) {
		node->right = insert_node(node->right, data);
	}
	return node;
}
main() {
	TreeNode* root = NULL;
	int value;
	while (scanf("%d", &value) != EOF) {
		root = insert_node(root, value);
	}
	postorder(root);
}

File: JAVA/28_Tree/BOJ_06_5639_이진검색트리.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;

class TreeNode {
    Integer data;
    TreeNode left, right;

    public TreeNode(Integer data) {
        this.data = data;
    }

    public void insert(int data) {
        if (this.data == null) this.data = data;
        if (data < this.data) {
            if (this.left == null) this.left = new TreeNode(data);
            else this.left.insert(data);
        } else if (data > this.data) {
            if (this.right == null) this.right = new TreeNode(data);
            else  this.right.insert(data);
        }
    }
}

public class Main {

    static StringBuilder sb = new StringBuilder();

    public static void postorder(TreeNode root) {
        if (root == null) return;
        if (root.left != null) postorder(root.left);
        if (root.right != null) postorder(root.right);
        sb.append(root.data).append("\n");
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        TreeNode root = new TreeNode(null);
        String data = null;
        while ((data = br.readLine()) != null) {
            root.insert(Integer.parseInt(data));
        }
        postorder(root);
        System.out.print(sb);
    }
}